### ISSUE TITLE
overload with decorator is ... unsound

### ISSUE THREAD

**Issue by KotlinIsland** [DetachHead/basedpyright#1463]
### Description

Code sample in [basedpyright playground](https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0DGZlAzi1AIIBctUfU1YsCgB9EaiQwxAChbEywADRQAHlwIAjAFbEGMAJRQAtAD4oAOTApiPfnagJWLegAEipCjVqDhwMGGk1TBQDYzMWGBB1ADpY13dyKjofKD8AoIiQQ1NgmBi42hcObyFU-0D1VHwAHyhM-Vt%2BEGIYAFcQFFV6ZpJKMhF4BGJpNP16ESgAXjKAgEYxoA)

```python
from typing import overload

class A:
    def __init__(self, x: object) -> None:
        pass

@overload
def foo(x: int) -> str: ...

@overload
def foo(x: str) -> int: ...

@A # no error
def foo(x: int | str):
    return x

reveal_type(foo) # Overload[...]

_ = foo(1)  # no error

```

not sure what the expected behaviour would be, if there should be an error on `@A`, or that `foo` should come out of the oven as `A`.....

**Comment by jorenham** [comment]
I think this should be identical to something like

```py
from typing import overload

class A:
    def __init__(self, x: object) -> None:
        pass

@overload
def __foo(x: int) -> str: ...
@overload
def __foo(x: str) -> int: ...
def __foo(x: int | str):
    return x

foo = A(__foo)
reveal_type(foo) # A
```

which seems to behave correctly

https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0DGZlAzi1AIIBctUfU1YsCgB9EaiQwxAChbEywADRQAHlwIAjAFbEGMAJRQAtAD4oAOTApiPfnagJWLegAEipCjVqDhY4GDBpNUwUA2MzFhgQdQA6ONo3EnIqOh9REX9A4MiQQ1MQmFj4tL8AoPVUfAAfKBz9W34QYhgAVxAUVXpMqABeTmlSsH1aJpJKMhF4BGJpTMMAYk4gA

so I think we've got ourselves a bugger

**Comment by Wizzerinus** [comment]
definitely should not error on `@A` itself because decorators don't have to return anything. Should make it so `@A` returns an A (then foo(1) would give an error).

EDIT: this is actually a good question on the semantics of `@overload`. I can definitely see why it can be useful to have `@overload` refer to undecorated version of the function, and can also see why it can be useful to refer to the decorated one... There is nothing in the typing spec about this, so we can probably go either way as long as it returns an error here

**Comment by DetachHead** [comment]
@jorenham an alternate way of thinking about it is that the `@A` decorator call is operating on the implementation signature:

```py
from typing import overload

class A:
    def __init__(self, x: object) -> None:
        pass

@overload
def foo(x: int) -> str: ...

@overload
def foo(x: str) -> int: ...

foo = A(1)
```

in which case i would expect the error to be on the overloads because they are overloading something that's not a function

**Comment by DetachHead** [comment]
> definitely should not error on `@A` itself because decorators don't have to return anything. Should make it so `@A` returns an A (then foo(1) would give an error).

not quite sure what you mean here. in this case the `@A` decorator call is returning something, an instance of `A`, but the issue i believe is that `A` has no `__call__` method.

however it looks like even if the class has a `__call__` method, pyright doesn't care if its signature is compatible with the overloads:

```py
from typing import overload

class A:
    def __init__(self, x: object) -> None:
        pass
    def __call__(self) -> int | str:
        ...

@overload
def foo(x: int) -> str: ...

@overload
def foo(x: str) -> int: ...

@A # should be an error because the implementation signature (A.__call__) isn't compatible with the overloads
def foo(x: int | str):
    return x
```

https://basedpyright.com/?typeCheckingMode=all&code=GYJw9gtgBALgngBwJYDsDmUkQWEMpgBuApiADZgCGAJgFC0DGZlAzi1AIIBctUfU1YsCgB9EaiQwxAChbEywADRQAHlwIAjAFbEGMAJRQAtAD4oAOTApiPfnagJWLXv0HCxDSmTIy5Cw6aYKPgAPlAsMCC29nwAdPH0AAJEpBQ0tG5QwGBg0mpBBsZmEVFQ8bFJKeRUdJnZufklAWaoMOrlSRxQAMThABZgAK5k1FAaxFCUKFCk4CBjupSDcrB9E1gIZMQQxMGUMEhW4UhoKPuDIBPSHLEeXj4ihkgsKADk%2BAyQjgcaW1AA7pI%2BqsJlU0tRnHUcnl1K0oGEmtE%2BJcYBdpiogA

**Comment by KotlinIsland** [comment]
so does an overload mean that it's applying to the final value of the impl, or the initial one ?

**Comment by jorenham** [comment]
Given that this is something bigger than basedpyright (or pyright for that matter), this probably something we should discuss in the discourse (yea I know...).

**Comment by KotlinIsland** [comment]
does a decorator on an overload impl mean that it is also applied to all signatures?

**Comment by Wizzerinus** [comment]
That is only explicitly stated on classmethod and staticmethod I believe: 

<img width="1030" height="304" alt="Image" src="https://github.com/user-attachments/assets/c01f162c-992e-489d-ab00-dc39ecf2f4e9" />

**Comment by jorenham** [comment]
for reference: https://typing.python.org/en/latest/spec/overload.html#invalid-overload-definitions

**Comment by jorenham** [comment]
I believe that `@deprecated` is the only one that applies to individual overloads (which shows that it would've been better as a `Deprecated[T]` return type, but I digress).

---

from the [spec](https://typing.python.org/en/latest/spec/directives.html#deprecated):

> With overloaded functions, the decorator may be applied to individual overloads, indicating that the particular overload is deprecated. The decorator may also be applied to the overload implementation function, indicating that the entire function is deprecated.
